Request data handling

talvarez0507 opened the pull request:
This PR implements the basic classes that will deal with the data for scheduling requests. These classes just hold events, tasks, start of day, end of day in one place.

talvarez0507 committed:
create class to model requests

talvarez0507 committed:
create java class to handle requests

hollyyuqizheng commented on portfolio/src/main/java/com/google/sps/data/ScheduleRequest.java:
```diff
@@ -0,0 +1,69 @@
+    this.endTime = Instant.parse(endTimeString);
+  }
+
+  /* Because all three fields are required, the following getters won't return null. */
```
nit: three => four fields

talvarez0507 replied:
Done.

hollyyuqizheng commented on portfolio/src/main/webapp/request.js:
```diff
@@ -0,0 +1,16 @@
+ * information that can be easily converted into a JSON string.
+ */
+class ScheduleRequest {
+  constructor(events, tasks, startTime, endTime) {
```
Would be nice to include comments on each parameter's type in the constructor.

talvarez0507 replied:
Done.

hollyyuqizheng commented on portfolio/src/main/java/com/google/sps/data/ScheduleRequest.java:
```diff
@@ -0,0 +1,69 @@
+
+import java.time.Instant;
+import java.util.Collection;
+/**
```
You might need to import CalendarEvent and Task?

talvarez0507 replied:
Not necessary because of the package.

talvarez0507 committed:
'three' to 'four'

hollyyuqizheng commented on portfolio/src/main/java/com/google/sps/data/ScheduleRequest.java:
```diff
@@ -0,0 +1,69 @@
+  private Collection<CalendarEvent> events;
+  private Collection<Task> tasks;
+  private final Instant startTime;
+  private final Instant endTime;
```
These times represent the start and end of working hours? If so, maybe a better name for the time variables could be "workHourStartTimeInstant"

talvarez0507 replied:
Done.

hollyyuqizheng reviewed (changes requested):
Some small suggestions

talvarez0507 committed:
startTime to workHoursStartTime and endTime

talvarez0507 committed:
fix names

talvarez0507 committed:
change task constructors and equals

talvarez0507 committed:
implement equals for TaskPriority

talvarez0507 committed:
make new methods that cast Instant to longs

talvarez0507 committed:
create TimeRange.java for dealing with ranges

This was taken directly from the week 5 tutorials to deal with ranges of time, the only change was that everything was cast o longs because seconds since epoch are large numbers

talvarez0507 committed:
fix param name in comments

talvarez0507 committed:
add algorithm class

talvarez0507 committed:
implement unit tests for scheduling algorithm

talvarez0507 committed:
implement servlet that schedules everything

talvarez0507 committed:
fix file extension

hollyyuqizheng commented on portfolio/src/main/java/com/google/sps/data/ScheduleRequest.java:
```diff
@@ -0,0 +1,72 @@
+   *     time. The string is in format Day Month Date Year HH:MM:SS GMT-Time-zone
+   * @param workHoursEndTimeString: a string representation of the work hours end time.
+   *     The string's format is the same as startTimeString's.
+   * All of these fields are required for a request (can be empty not null).
```
When will these fields be empty?

talvarez0507 replied:
Done.

hollyyuqizheng commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+    // meeting. As events are processed, this changes to their end times.
+    long earliestPossibleSoFar = startTimeMinutes;
+    for (CalendarEvent event : events) {
+      // Make sure that there is some time between the events and is it not
```
nit: "is it" => "it is"

talvarez0507 replied:
Done.

hollyyuqizheng commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+   * determines should be scheduled for the person.
+   */
+  public Collection<Task> greedy(Collection<CalendarEvent> events, Collection<Task> tasks,
+      long startTimeMinutes, long endTimeMinutes) {
```
I would change the variable names to something like workHourStartTimeMinutes, just as you did in other classes, to keep things uniform.

talvarez0507 replied:
Done.

hollyyuqizheng commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+    int j = 0;
+    // Integer indicating the time (in minutes) we are currently trying to
+    // schedule events in
+    long time = 0;
```
I would rename this to currentTime or something similar.

talvarez0507 replied:
Done.

hollyyuqizheng reviewed (changes requested):
Some suggestions on style mostly.

talvarez0507 committed:
update findschedule

talvarez0507 committed:
parse all the scheduling data

talvarez0507 committed:
clean file for style

talvarez0507 committed:
implement new larger test

hollyyuqizheng commented on portfolio/src/test/java/com/google/sps/FindScheduleTest.java:
```diff
@@ -0,0 +1,136 @@
+  // TODO(tomasalvarez): Test edge cases
+  // 1: Events out of the scheduling bounds
+  // 2: Scheduling time is more than 24 hours
+  // 3: ???
```
Maybe we could do something like only half of the tasks can be scheduled because there isn't enough free space for the day?

talvarez0507 replied:
I added a new TODO for this case, thank you for the suggestion.

raulcruise commented on portfolio/src/test/java/com/google/sps/FindScheduleTest.java:
```diff
@@ -0,0 +1,100 @@
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** */
```
Remove line or add a comment.

talvarez0507 replied:
Done.

talvarez0507 committed:
remove unused Optional import

talvarez0507 committed:
rename work hours variables to match other files

talvarez0507 committed:
rename time variable to be more descriptive

talvarez0507 committed:
fix grammar nit

talvarez0507 committed:
update parameter description

talvarez0507 committed:
include parameter descriptions

talvarez0507 committed:
remove blank comment

talvarez0507 committed:
include new TODO

hollyyuqizheng commented on portfolio/src/main/java/com/google/sps/data/TimeRange.java:
```diff
@@ -0,0 +1,79 @@
+    return other instanceof TimeRange && equals(this, (TimeRange) other);
+  }
+
+  @Override
```
I feel like we might end up not needing this hashCode function, so maybe it's okay to delete?

talvarez0507 replied:
Done.

raulcruise commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+    int j = 0;
+    // Integer indicating the start time (in minutes) we are currently trying to
+    // schedule events in
+    long scheduleTimeSeconds = 0;
```
The comment indicates that you're declaring a int in units "minutes" but you end up declaring a long int with the unit seconds.

talvarez0507 replied:
Done.

raulcruise commented on portfolio/src/main/java/com/google/sps/data/ScheduleRequest.java:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  // These methods return the instant as a number of seconds from 00:00:00 UTC
+  // on 1 January 1970
```
Nit: period at end.

talvarez0507 replied:
Done.

raulcruise commented on portfolio/src/main/webapp/request.js:
```diff
@@ -0,0 +1,25 @@
+ * information that can be easily converted into a JSON string.
+ */
+class ScheduleRequest {
+    /**
```
Nit: Misaligned beginning to javadoc style format.

talvarez0507 replied:
Done.

talvarez0507 committed:
missing period

talvarez0507 committed:
fix javadoc comment

talvarez0507 committed:
fix comment

rgcase commented on portfolio/src/main/java/com/google/sps/data/CalendarEvent.java:
```diff
@@ -47,4 +47,12 @@ public Instant getStartTimeInstant() {
     return endTime;
   }
+
+  public long getStartTimeLong() {
```
Here and below, I don't think the `Long` suffix is useful here. What would be more useful is if the suffix was the units instead, like `getStartTimeSeconds()`.

talvarez0507 replied:
I removed the methods.

rgcase commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+      Comparator.comparing(Task::getDurationSeconds).thenComparing(Task::getName);
+
+  /**
+   * This method returns a Task Collection which the greedy algorithm
```
nit: The "This method..." is redundant, the reader knows that the comment refers to this method.

Also, it's not really obvious to me from the comment + method signature what this method is supposed to do. It looks like it's using a greedy algorithm to choose where each task should be scheduled based on existing calendar events (and working hours), but it just returns `Task`s which don't have a start and end time, so it doesn't seem to be actually choosing times for these. Does this just filter out ones that are impossible to schedule?

talvarez0507 replied:
I implemented the ScheduledTask wrapper you suggested so now it returns a collection of those instead.

rgcase commented on portfolio/src/main/java/com/google/sps/data/Task.java:
```diff
@@ -13,32 +14,55 @@
+   * field as a result of getting scheduled by the scheduling algorithm.
    */
-  public Task(String name, String description, Duration duration, TaskPriority priority) {
+  public Task(String name, String description, long durationMinutes, TaskPriority priority) {
```
I think we should try to stick with using `Duration` here. Why do we need to use `long` instead?

talvarez0507 replied:
I reverted back to Duration.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/CalendarEvent.java:
```diff
@@ -47,4 +47,12 @@ public Instant getStartTimeInstant() {
     return endTime;
   }
+
+  public long getStartTimeLong() {
```
Consider clarifying the time unit in the method name.

ex: getStartTimeSecondsLong (or just getStartTimeSeconds)

jamesnam-google replied:
By the way, do we really need this method? I think the callers can choose to invoke getEpochSecond themselves. I think that is one major benefit of returning Instant; the callers can decide how to work with the instant.

talvarez0507 replied:
I removed the methods.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/CalendarEvent.java:
```diff
@@ -47,4 +47,12 @@ public Instant getStartTimeInstant() {
+    return startTime.getEpochSecond();
+  }
+
+  public long getEndTimeLong() {
```
Same here

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+import java.util.Comparator;
+import java.util.List;
+
+public final class FindSchedule {
```
Consider adding some class comment. Also, how do you think we can make this class extensible? Consider adding TODO comment if you think it can be made more extensible, as we add scheduling algorithms.

jamesnam-google replied:
Also does this class have any instance variable? Maybe it can be a static class, as well as all of its methods.

talvarez0507 replied:
I left a TODO to refactor it into an abstract class for the next PR and made it static.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+import java.util.List;
+
+public final class FindSchedule {
+  private static final Comparator<CalendarEvent> sortAscending =
```
I think the variable name should make it clear which field is used for sorting.

ex:
sortByEventStartTimeAscending

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+  private static final Comparator<CalendarEvent> sortAscending =
+      Comparator.comparingLong(CalendarEvent::getStartTimeLong);
+
+  private static final Comparator<Task> sortByDuration =
```
I think the variable name should make it clear that it uses task duration, then name.

ex:
sortByTaskDurationThenName

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+   * This method returns a Task Collection which the greedy algorithm
+   * determines should be scheduled for the person.
+   */
+  public Collection<Task> greedy(Collection<CalendarEvent> events, Collection<Task> tasks,
```
Let's try to have one param per line.

ex:
greedy(
    Collection..,
    Collection...,
    ....,
    ...);

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+   * determines should be scheduled for the person.
+   */
+  public Collection<Task> greedy(Collection<CalendarEvent> events, Collection<Task> tasks,
+      long workHoursStartTimeSeconds, long workHoursEndTimeSeconds) {
```
I think we should try to use Instant as parameter wherever possible. Can this method take in Instant params?

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+      Comparator.comparing(Task::getDurationSeconds).thenComparing(Task::getName);
+
+  /**
+   * This method returns a Task Collection which the greedy algorithm
```
Consider clarifying further in the method comment what you mean by greedy algorithm. For example does it try to fit in the task with the highest priority first?

talvarez0507 replied:
Changed method name to be shortestTaskFirst and updated comment.

rgcase commented on portfolio/src/main/java/com/google/sps/data/Task.java:
```diff
@@ -13,32 +14,55 @@
   private final Optional<String> description;
   private final Duration duration;
   private final TaskPriority priority;
+  private Instant scheduledTime;
```
I don't really like how a `Task` will sometimes have a `scheduledTime` and sometimes not, and there's not really any way to know about it ahead of time. Consider instead creating a wrapper class like:
```
public class ScheduledTask {
  private final Task task;
  private final Instant startTime;

  ...
}
```

then you can statically know whether a `Task` has been scheduled or not.

talvarez0507 replied:
I implemented ScheduledTask.

rgcase commented on portfolio/src/main/java/com/google/sps/data/Task.java:
```diff
@@ -53,7 +77,33 @@ public Duration getDuration() {
+  }
+
+  private static boolean equals(Task a, Task b) {
+    return a.name.equals(b.name) && a.description.equals(b.description)
```
nit: Line break before each `&&`, don't mix lines with one clause and lines with multiple clauses.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+   * working hours.
+   */
+  public static TimeRange[] emptyTimeRanges(
+      List<CalendarEvent> events, long workHoursStartTimeSeconds, long workHoursEndTimeSeconds) {
```
Can we try to take in Instant objects rather than long?

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+   * time that are empty of events and lie completely inside the person's
+   * working hours.
+   */
+  public static TimeRange[] emptyTimeRanges(
```
1. Is there a reason why you return array instead of list? Lists are more flexible and easier to work with. Consider returning a list of TimeRange.
2. I think it'd be nice to clarify in the method name that it gets or computes the time ranges. So for example, computeEmptyTimeRanges or getEmptyTimeRanges.

talvarez0507 replied:
Changed it to return a List instead and renamed it to getEmptyTimeRanges.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+    List<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    // This represents the earliest time that we can schedule a window for the
+    // meeting. As events are processed, this changes to their end times.
+    long earliestPossibleSoFar = workHoursStartTimeSeconds;
```
The time unit is not clear in the variable name. And consider using Instant here as well.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+    for (CalendarEvent event : events) {
+      // Make sure that there is some time between the events and it is not
+      // later than the person's working hours ending time.
+      if (event.getStartTimeLong() - earliestPossibleSoFar > (long) 0
```
event.getStartTimeLong() - earliestPossibleSoFar > (long) 0

is equivalent to

event.getStartTimeLong() > earliestPossibleSoFar

talvarez0507 replied:
Done.

rgcase commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+      // The task can be scheduled in the current time range.
+      if (scheduleTimeSeconds + task.getDurationSeconds() <= timeRange.end()) {
+        Task scheduledTask;
+        if (task.getDescription().isPresent()) {
```
This is kind of messy, but I think it'll go away if you start using a new `ScheduledTask` like I mentioned in a previous comment.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+      earliestPossibleSoFar = Math.max(earliestPossibleSoFar, event.getEndTimeLong());
+    }
+    // The end of the work hours is potentially never included so we check.
+    if (workHoursEndTimeSeconds - earliestPossibleSoFar > 0) {
```
Equivalent to,

workHoursEndTimeSeconds > earliestPossibleSoFar

talvarez0507 replied:
Done.

rgcase commented on portfolio/src/main/java/com/google/sps/data/TimeRange.java:
```diff
@@ -0,0 +1,79 @@
+ * providing methods to make ranges easier to work with (e.g. {@code overlaps}).
+ */
+public final class TimeRange {
+  private final long start;
```
Can we use `Instant` here instead of `long`?

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+  public Collection<Task> greedy(Collection<CalendarEvent> events, Collection<Task> tasks,
+      long workHoursStartTimeSeconds, long workHoursEndTimeSeconds) {
+    List<CalendarEvent> eventsList = new ArrayList<CalendarEvent>(events);
+    Task[] tasksArray = tasks.toArray(new Task[tasks.size()]);
```
As I wrote in my comment below, consider using list instead of array, unless there is a strong reason to use array.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+    Task[] tasksArray = tasks.toArray(new Task[tasks.size()]);
+    Collections.sort(eventsList, sortAscending);
+    Arrays.sort(tasksArray, 0, tasks.size(), sortByDuration);
+    TimeRange[] availableTimes =
```
Similar comment here about list vs array.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+    int j = 0;
+    // Integer indicating the start time (in seconds) we are currently trying to
+    // schedule events in
+    long scheduleTimeSeconds = 0;
```
Consider using Instant wherever possible.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+    // later ones will fit either). We create new Task objects for the result
+    // so data structures passed in are never changed.
+    while (i < availableTimes.length && j < tasksArray.length) {
+      TimeRange timeRange = availableTimes[i];
```
Consider renaming to availableTimeRange

talvarez0507 replied:
Done.

rgcase commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -0,0 +1,93 @@
+
+  private static Collection<CalendarEvent> collectEventsFromJson(JSONArray eventsArray) {
+    Collection<CalendarEvent> events = new ArrayList<CalendarEvent>();
+    for (int i = 0; i < eventsArray.length(); i++) {
```
I think you should be able to use `for (JSONObject eventJsonObject : eventsArray) { ... }` rather than indexing into the `JSONArray`.

talvarez0507 replied:
org.json.simple.JSONArray implements a raw Iterator so I changed it to a for each loop but still had to cast each Object into a JSONObject.

rgcase commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -0,0 +1,93 @@
+    Collection<CalendarEvent> events = new ArrayList<CalendarEvent>();
+    for (int i = 0; i < eventsArray.length(); i++) {
+      JSONObject eventAsJSONObject = eventsArray.getJSONObject(i);
+      String name = (String) eventAsJSONObject.get("name");
```
I think there are built in methods like `.getString(...)` that you can use here. Here is the API I found https://stleary.github.io/JSON-java/org/json/JSONArray.html

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+      // The task can be scheduled in the current time range.
+      if (scheduleTimeSeconds + task.getDurationSeconds() <= timeRange.end()) {
+        Task scheduledTask;
+        if (task.getDescription().isPresent()) {
```
It would be simpler to use ternary operator in the constructor.

ex:
Task scheduledTask = new Task(
    task.getName(),
    task.getDescription().isPresent() ? task....get() : null,
    ...);

talvarez0507 replied:
Refactored using a new class so this wasn't necessary.

rgcase reviewed (commented):
Top-level comment: I think we should be able to change all the `long`s to `Instant` or `Duration`.  It'll make it much safer doing computations on these values and less likely to introduce a bug later.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+    int j = 0;
+    // Integer indicating the start time (in seconds) we are currently trying to
+    // schedule events in
+    long scheduleTimeSeconds = 0;
```
Consider renaming to currentScheduleTimeSeconds

talvarez0507 replied:
Renamed to currentScheduleTime because I refactored it into an Instant.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/ScheduleRequest.java:
```diff
@@ -0,0 +1,75 @@
+
+import java.time.Instant;
+import java.util.Collection;
+/**
```
Line break above this line. Also, can you run Java autoformatter and ESLINT on the code change?

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/ScheduleRequest.java:
```diff
@@ -0,0 +1,75 @@
+  // These methods return the instant as a number of seconds from 00:00:00 UTC
+  // on 1 January 1970.
+
+  public long getWorkHoursStartTimeLong() {
```
As I suggested earlier, I think this class can just have methods that return Instant. Then the callers can decide which methods to call on the Instant (like getEpochSecond).

talvarez0507 replied:
Deleted the methods.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/ScheduleRequest.java:
```diff
@@ -0,0 +1,75 @@
+  public Collection<Task> getTasks() {
+    return tasks;
+  }
+  public Instant getWorkHoursStartTimeInstant() {
```
Line break between methods

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/ScheduleRequest.java:
```diff
@@ -0,0 +1,75 @@
+   * schedule for that day or the user hits start scheduling before they add any tasks
+   * (maybe by accident).
+   */
+  public ScheduleRequest(Collection<CalendarEvent> events, Collection<Task> tasks,
```
Let's have one param per line.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/Task.java:
```diff
@@ -13,32 +14,55 @@
+   * The constructor allows for Tasks to have their scheduled time be set. New tasks being made
+   * for scheduling purposes use this constructor instead.
+   */
+  public Task(String name, String description, long durationMinutes, TaskPriority priority,
```
Let's have one param per line

talvarez0507 replied:
I changed some names and it fit on one line again.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/Task.java:
```diff
@@ -53,7 +77,33 @@ public Duration getDuration() {
     return priority.getPriority();
   }
+
+  public long getDurationSeconds() {
```
Consider removing this method, because the caller can get the duration and invoke getSeconds()

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -0,0 +1,93 @@
+
+    ScheduleRequest scheduleRequest =
+        new ScheduleRequest(events, tasks, workHoursStartTimeString, workHoursEndTimeString);
+    FindSchedule findSchedule = new FindSchedule();
```
I think FindSchedule can be static. Then you don't have to instantiate it, and just call

FindSchedule.greedy(...)

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -0,0 +1,93 @@
+        new ScheduleRequest(events, tasks, workHoursStartTimeString, workHoursEndTimeString);
+    FindSchedule findSchedule = new FindSchedule();
+    Collection<Task> scheduledTasks =
+        findSchedule.greedy(scheduleRequest.getEvents(), scheduleRequest.getTasks(),
```
Consider having one param per line.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -0,0 +1,93 @@
+            scheduleRequest.getWorkHoursStartTimeLong(), scheduleRequest.getWorkHoursEndTimeLong());
+
+    Gson gson = new Gson();
+    String resultJSON = gson.toJson(scheduledTasks);
```
Typically variable names have camel-case, including acronyms.

ex:
resultJson

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -0,0 +1,93 @@
+    response.getWriter().println(resultJSON);
+  }
+
+  private static Collection<CalendarEvent> collectEventsFromJson(JSONArray eventsArray) {
```
Consider renaming to, collectEventsFromJsonArray

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -0,0 +1,93 @@
+    return events;
+  }
+
+  private static Collection<Task> collectTasksFromJson(JSONArray tasksArray) {
```
Similar comment here about method name

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -0,0 +1,93 @@
+      JSONObject taskAsJSONArray = tasksArray.getJSONObject(i);
+      String name = (String) taskAsJSONArray.get("name");
+      String description = (String) taskAsJSONArray.get("description");
+      long duration = Long.valueOf((int) taskAsJSONArray.get("durationMinutes"));
```
duration => durationMinutes

That being said, I would encourage that you perform the conversion to Duration as soon as possible, then pass it as a param to Task constructor rather than passing minutes. Same with Instant.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,116 @@
+        Task scheduledTask;
+        if (task.getDescription().isPresent()) {
+          scheduledTask =
+              new Task(task.getName(), task.getDescription().get(), task.getDuration().toMinutes(),
```
Consider having one param per line, here and below.

talvarez0507 replied:
It now fits on one line because of the refactoring.

talvarez0507 committed:
implement change requests

talvarez0507 committed:
implement change requests

talvarez0507 committed:
implement change requests

talvarez0507 committed:
implement change requests

talvarez0507 committed:
implement change requests

talvarez0507 committed:
implement change requests

talvarez0507 committed:
implement new ScheduledTask class

talvarez0507 committed:
remove tests that no longer work

These tests were based on the previous classes with different constructors so most of them do not work. I left a TODO that I will fix them tomorrow.

talvarez0507 committed:
update CalendarEvent constructor to take Instant

talvarez0507 committed:
use the common version of this file

talvarez0507 committed:
fix field names

talvarez0507 committed:
fix variable names

talvarez0507 committed:
update tests to reflect class changes

talvarez0507 committed:
There is no conversion so this test is outdated

talvarez0507 committed:
change loops to for each loops

rgcase commented on portfolio/src/main/java/com/google/sps/data/CalendarEvent.java:
```diff
@@ -5,34 +5,32 @@
   private String name;
-  private final Instant startTime;
-  private final Instant endTime;
+  private final Instant startTimeInstant;
```
I don't think you need the `*Instant` suffix anywhere in this class, the type is right there.

talvarez0507 replied:
Done.

rgcase commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,118 @@
+    int i = 0;
+    // Index of Task we are trying to schedule.
+    int j = 0;
+    // Integer indicating the start time (in minutes) we are currently trying to
```
nit: The comment still says "Integer" instead of "Instant". You can remove the "(in minutes)" after making the change too.

talvarez0507 replied:
Done.

rgcase commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,118 @@
+
+    List<ScheduledTask> scheduledTasks = new ArrayList<ScheduledTask>();
+    // Index of TimeRange we are in.
+    int i = 0;
```
Rather than have these comments, maybe it's more clear if you just name the variables appropriately, like `rangeIndex` and `taskIndex` or something like that. Then when you're reading the code below you don't have to look back up here all the time to find out which is which.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/test/java/com/google/sps/FindScheduleTest.java:
```diff
@@ -0,0 +1,130 @@
+
+  // TODO(tomasalvarez): rewrite all these tests to reflect changes made 
+
+  /*
```
Can we have unit test that goes with the code change made? Also, please do not have commented out code here like this. It is better to remove from the code-base (you can always refer to the pull request to add it back).

talvarez0507 replied:
I rewrote the tests after I implemented the abstract class you suggested. It feels counter-intuitive to fix this now since I already rewrote the tests based on that implementation. I will remove the file from this PR and include all of the changes in the next.

jamesnam-google commented on portfolio/src/test/java/com/google/sps/FindScheduleTest.java:
```diff
@@ -0,0 +1,130 @@
+  private static Instant TIME_1700 = BEGINNING_OF_DAY.plusSeconds(17 * TIME_60_MINUTES);
+  private static Instant TIME_1800 = BEGINNING_OF_DAY.plusSeconds(18 * TIME_60_MINUTES);
+
+
```
Unnecessary line break

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/webapp/request.js:
```diff
@@ -0,0 +1,25 @@
+   * @param endTime: a string representation of the work hours end time.
+   *     The string's format is the same as startTimeString's.
+   */
+  constructor(events, tasks, startTime, endTime) {
```
Can this method take in the Date Javascript object, rather than a String?

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/ScheduleRequest.java:
```diff
@@ -0,0 +1,68 @@
+  public ScheduleRequest(
+      Collection<CalendarEvent> events,
+      Collection<Task> tasks,
+      String workHoursStartTimeString,
```
Is it possible for this method to take in Instant for start time and end time?

talvarez0507 replied:
Since the servlet handles all of the unpacking at this point, I just removed this file.

rgcase commented on portfolio/src/test/java/com/google/sps/FindScheduleTest.java:
```diff
@@ -0,0 +1,130 @@
+    Assert.assertEquals(expected, actual);
+  }
+
+  // TODO(tomasalvarez): rewrite all these tests to reflect changes made 
```
Do you plan to fix these tests in this PR?

talvarez0507 replied:
I rewrote the tests after I refactored a lot of the code. I will remove the file from this PR and on the next PR the updated files will include all of the changes.

talvarez0507 committed:
fix javadoc comment

talvarez0507 committed:
remove Instant suffix everywhere

talvarez0507 committed:
delete class

This class was mostly meant to pack unpack the data from the UI but now that the servlet exists this class is not useful.

talvarez0507 committed:
update removal of ScheduleRequest class

talvarez0507 committed:
implement change requests

Fix comments that were outdated, change variable name, remove some extra blank lines.

talvarez0507 committed:
delete outdated file

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/Task.java:
```diff
@@ -53,7 +52,22 @@ public Duration getDuration() {
+    return other instanceof Task && equals(this, (Task) other);
+  }
+
+  private static boolean equals(Task a, Task b) {
```
I think you'll want to start using task ID soon. The current approach won't work well when you begin to integrate with Tasks API.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/TimeRange.java:
```diff
@@ -0,0 +1,73 @@
+  /**
+   * Creates a {@code TimeRange} from {@code start} to {@code end}. Whether or not {@code end} is
+   * included in the range will depend on {@code inclusive}. If {@code inclusive} is {@code true},
+   * then @{code end} will be in the range.
```
{@code end} instead of @{code end}

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/webapp/request.js:
```diff
@@ -0,0 +1,23 @@
+class ScheduleRequest {
+  /**
+   * All fields are required for this object.
+   * @param events: events happening the day of the scheduling,
```
Consider clarifying the type of events and tasks. Are they Javascript arrays of * ?

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,112 @@
+      // Make sure that there is some time between the events and it is not
+      // later than the person's working hours ending time.
+      if (event.getStartTimeInstant().isAfter(earliestNonScheduledInstant)
+          && (!event.getStartTimeInstant().isAfter(endTime))) {
```
Unnecessary parents around !event....

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,112 @@
+// TODO(tomasalvarez): Refactor this class into an abstract class where each of
+// the scheduling algorithms can be their own class that extends this one.
+
+public class FindSchedule {
```
Consider adding TODO comment to add test coverage.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/FindSchedule.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+  /**
+   * This method returns an TimeRange array which represent the periods of time that are empty of
```
array => list. Arrays and Lists are different concepts in Java.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -0,0 +1,92 @@
+    response.getWriter().println(resultJson);
+  }
+
+  private static Collection<CalendarEvent> collectEventsFromJsonArray(JSONArray eventsArray) {
```
Consider adding TODO comment to add test for this method and the one below.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/TimeRange.java:
```diff
@@ -0,0 +1,73 @@
+ * Class representing a span of time, enforcing properties (e.g. start comes before end) and
+ * providing methods to make ranges easier to work with (e.g. {@code overlaps}).
+ */
+public final class TimeRange {
```
Consider adding TODO comment to add test for this class.

talvarez0507 replied:
Done.

jamesnam-google reviewed (approved):
Approving the change with comments

talvarez0507 committed:
fix comment and add TODO

talvarez0507 committed:
update param comments

talvarez0507 committed:
implement change requests

fix comment, remove extra parenthesis

talvarez0507 committed:
include TODO for new tests

talvarez0507 committed:
change equals() to public

talvarez0507 committed:
update to new class getters

talvarez0507 committed:
Merge branch 'master' into tomas
